examples/matching_engine/src/contract.rs

```diff
@@ -0,0 +1,640 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#![cfg_attr(target_arch = "wasm32", no_main)]
+
+mod state;
+
+use matching_engine::{
+    product_price_amount, ApplicationCall, KeyBook, Message, ModifyCount, Operation, Order,
+    OrderEntry, OrderId, OrderNature, Price, Transfer,
+};
+use state::{LevelView, MatchingEngine, MatchingEngineError};
+use std::cmp::min;
+
+use async_trait::async_trait;
+use fungible::{Account, AccountOwner, Destination, FungibleTokenAbi};
+use linera_sdk::{
+    base::{Amount, ApplicationId, SessionId, WithContractAbi},
+    contract::system_api,
+    ensure, ApplicationCallResult, CalleeContext, Contract, ExecutionResult, MessageContext,
+    OperationContext, SessionCallResult, ViewStateStorage,
+};
+
+linera_sdk::contract!(MatchingEngine);
+
+impl WithContractAbi for MatchingEngine {
+    type Abi = matching_engine::MatchingEngineAbi;
+}
+
+#[async_trait]
+impl Contract for MatchingEngine {
+    type Error = MatchingEngineError;
+    type Storage = ViewStateStorage<Self>;
+
+    async fn initialize(
+        &mut self,
+        _context: &OperationContext,
+        _argument: (),
+    ) -> Result<ExecutionResult<Self::Message>, Self::Error> {
+        Ok(ExecutionResult::default())
+    }
+
+    async fn execute_operation(
+        &mut self,
+        context: &OperationContext,
+        operation: Operation,
+    ) -> Result<ExecutionResult<Self::Message>, Self::Error> {
+        let mut result = ExecutionResult::default();
+        match operation {
+            Operation::Order { order } => {
+                if context.chain_id == system_api::current_application_id().creation.chain_id {
+                    self.execute_order_local(order).await?;
+                } else {
+                    self.execute_order_remote(&mut result, order).await?;
+                }
+            }
+        }
+        Ok(result)
+    }
+
+    async fn execute_message(
+        &mut self,
+        context: &MessageContext,
+        message: Message,
+    ) -> Result<ExecutionResult<Self::Message>, Self::Error> {
+        ensure!(
+            context.chain_id == system_api::current_application_id().creation.chain_id,
+            Self::Error::AdminChainOnly
+        );
+        match message {
+            Message::Order { order } => {
+                self.execute_order_local(order).await?;
+            }
+        }
+        Ok(ExecutionResult::default())
+    }
+
+    async fn handle_application_call(
+        &mut self,
+        _context: &CalleeContext,
+        argument: ApplicationCall,
+        _sessions: Vec<SessionId>,
+    ) -> Result<ApplicationCallResult<Self::Message, Self::Response, Self::SessionState>, Self::Error>
+    {
+        let mut result = ApplicationCallResult::default();
+        match argument {
+            ApplicationCall::Order { order } => {
+                self.execute_order_remote(&mut result.execution_result, order)
+                    .await?;
```

Comment:
I will readily admit that I do not completely answer the process.
The `handle_application_call` are creating the `ApplicationCallResult`. If I used the `execute_order_local` then I am not creating the ApplicationCallResult and just returning the default.

I would think that the application calls have to be locals since we need to install the application locally before creating it.

So, it is really mysterious to me and since it was not critical for the application then I must have written some stupid things. I would appreciate help on this.

Reply:
My understanding is that if the matching engine supports being called by other applications, it should probably support being called on the main chain (where the application was created), too. Or at least check for it and return an error.

But maybe it doesn't need to support application calls at all, at least for now? I.e. could we just set `type ApplicationCall = ()` and always return an error here?

Reply:
Crowd-funding campaigns and order books should be similar. See also comments below.

Reply:
I coded the `handle_application_call` in a similar way to `execute_operation`. Let me know what you think.